Dockerfile

```diff
@@ -0,0 +1,164 @@
+WORKDIR $UNDERLAY_WS
+
+# install underlay dependencies
+RUN . /opt/ros/$ROS_DISTRO/setup.sh && \
+    apt-get update && rosdep install -q -y \
+      --from-paths src \
+      --ignore-src \
+    && rm -rf /var/lib/apt/lists/*
+
+# copy underlay source
+COPY --from=cache $UNDERLAY_WS ./
+
+# build underlay source
+ARG UNDERLAY_MIXINS="release"
+RUN . /opt/ros/$ROS_DISTRO/setup.sh && \
+    colcon build \
+      --symlink-install \
+      --mixin $UNDERLAY_MIXINS \
+      --cmake-args \
+        --no-warn-unused-cli \
+        -DGTSAM_BUILD_TESTS=OFF \
+        -DGTSAM_BUILD_EXAMPLES_ALWAYS=OFF \
+        -DGTSAM_BUILD_UNSTABLE=ON \
+        -DGTSAM_POSE3_EXPMAP=ON \
+        -DGTSAM_ROT3_EXPMAP=ON \
+        -DOPENCV_EXTRA_MODULES_PATH=$UNDERLAY_WS/src/opencv/opencv/contrib/modules
+      # --event-handlers console_direct+
+
+# copy overlay manifests
+ENV OVERLAY_WS /opt/overlay_ws
+COPY --from=cache /tmp/overlay_ws $OVERLAY_WS
+WORKDIR $OVERLAY_WS
+
+# install overlay dependencies
+RUN . $UNDERLAY_WS/install/setup.sh && \
+    apt-get update && rosdep install -q -y \
+      --from-paths \
+        src \
+        $UNDERLAY_WS/src \
+      --ignore-src \
+    && rm -rf /var/lib/apt/lists/*
+
+# copy overlay source
+COPY --from=cache $OVERLAY_WS ./
+
+# build overlay source
+ARG OVERLAY_MIXINS="release ccache"
+RUN . $UNDERLAY_WS/install/setup.sh && \
+    colcon build \
+      --symlink-install \
+      --mixin $OVERLAY_MIXINS \
+      --cmake-args \
+        --no-warn-unused-cli \
+        -DCMAKE_CXX_FLAGS="\
+          -Wno-parentheses \
+          -Wno-reorder \
+          -Wno-sign-compare \
+          -Wno-unused-but-set-variable \
+          -Wno-unused-function \
+          -Wno-unused-parameter \
+          -Wno-unused-value \
+          -Wno-unused-variable"
+      # --event-handlers console_direct+
+
+# copy ros manifests
+ENV ROS_WS /opt/ros_ws
+COPY --from=cache /tmp/ros_ws $ROS_WS
+WORKDIR $ROS_WS
+
+# install overlay dependencies
+RUN . $OVERLAY_WS/install/setup.sh && \
+    apt-get update && rosdep install -q -y \
+      --from-paths \
+        src \
+        $UNDERLAY_WS/src \
+        $OVERLAY_WS/src \
+      --ignore-src \
+      --skip-keys "\
+        Boost" \
+    && rm -rf /var/lib/apt/lists/*
+
+# copy overlay source
+COPY --from=cache $ROS_WS ./
+
+# build overlay source
+ARG ROS_MIXINS="release ccache"
+RUN . $OVERLAY_WS/install/setup.sh && \
+    colcon build \
+      --symlink-install \
+      --mixin $ROS_MIXINS \
+      --cmake-args \
+        --no-warn-unused-cli \
+        -DCMAKE_CXX_FLAGS="\
+          -Wno-sign-compare \
+          -Wno-unused-value \
+          -Wno-unused-variable \
+          -Wno-unused-but-set-variable \
+          -Wno-reorder \
+          -Wno-parentheses \
+          -Wno-unused-parameter"
+      # --event-handlers console_direct+
+
+# source overlay from entrypoint
+RUN sed --in-place \
+      's|^source .*|source "$ROS_WS/install/setup.bash"|' \
+      /ros_entrypoint.sh
```

Comment:
Isn't this file in another pr?

Reply:
This PR is still labeled WIP, so I haven't refactored everything possible into separate PRs yet. This is also used for development of the PR, so it was handy for me to keep in branch.